src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,285 @@
+    SplitIterMany(ParquetFileSplitter<'a>),
+}
+
+/// Iterator that performs a split on the provided [`ScanTaskRef`]
+struct ParquetFileSplitter<'a> {
+    // Iterator state that tracks iteration progress
+    current_idx: usize,
+    accumulated_rowgroup_ids: Vec<usize>,
+    estimated_accumulated_materialized_size: f64,
+
+    // Cached items
+    scan_task: ScanTaskRef,
+    file_metadata: FileMetaData,
+    config: &'a DaftExecutionConfig,
+
+    // Call once and cache this, since it can be expensive when called in a loop
+    scan_task_in_memory_size_estimate: Option<usize>,
+}
+
+impl SplittableScanTaskRef {
+    fn into_iter(self, cfg: &DaftExecutionConfig) -> SplittableScanTaskRefIterator {
+        match self {
+            Self::NoSplit(st) => SplittableScanTaskRefIterator::NoSplitIterOnce(st, false),
+            Self::Split(st, fm) => {
+                let scan_task_in_memory_size_estimate = st.estimate_in_memory_size_bytes(Some(cfg));
+                SplittableScanTaskRefIterator::SplitIterMany(ParquetFileSplitter {
+                    current_idx: 0,
+                    accumulated_rowgroup_ids: Vec::new(),
+                    estimated_accumulated_materialized_size: 0.,
+                    scan_task: st,
+                    file_metadata: fm,
+                    config: cfg,
+                    scan_task_in_memory_size_estimate,
+                })
+            }
+        }
+    }
+}
+
+impl<'a> Iterator for SplittableScanTaskRefIterator<'a> {
+    type Item = ScanTaskRef;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::NoSplitIterOnce(st, has_returned) => {
+                if !*has_returned {
+                    *has_returned = true;
+                    Some(st.clone())
+                } else {
+                    None
+                }
+            }
+            Self::SplitIterMany(splitter) => splitter.next(),
+        }
+    }
+}
+
+impl<'a> Iterator for ParquetFileSplitter<'a> {
+    type Item = ScanTaskRef;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while !self.exhausted() {
+            let emitted_result = self.advance_state();
+            if let Some(emitted_result) = emitted_result {
+                return Some(emitted_result);
+            }
+        }
+        None
+    }
+}
+
+impl<'a> ParquetFileSplitter<'a> {
+    /// Returns whether or not this Splitter is done producing items
+    pub fn exhausted(&self) -> bool {
+        self.accumulated_rowgroup_ids.is_empty()
+            && self.current_idx >= self.file_metadata.row_groups.len()
+    }
+
+    /// Advances the state, optionally emitting a ScanTaskRef
+    pub fn advance_state(&mut self) -> Option<ScanTaskRef> {
+        if self.current_idx < self.file_metadata.row_groups.len() {
+            self.accumulate()
+        } else {
+            self.flush()
+        }
+    }
+
+    fn update_accumulator(&mut self, new_rg_idx: usize, rg_estimated_in_memory_size_bytes: f64) {
```

Comment:
these function names aren't very descriptive tbh

Reply:
Renamed and added some docstrings. I can also further refactor to pull it out into a separate `State` struct if you'd prefer that, but I felt like that was a little too verbose